cmd/snap/main.go

```diff
@@ -473,6 +474,32 @@ func main() {
 			fmt.Fprintf(Stderr, i18n.G("cannot resolve snap app %q: %v"), snapApp, err)
 			os.Exit(46)
 		}
+
+		// If running a desktop file from a confined snap process, then run via the privileged launcher.
+		// Each desktop file generated by snapd sets the BAMF_DESKTOP_FILE_HINT environment variable in
+		// the Exec line, so checking for this indicates this application is being run via a desktop file and not directly.
+		snapPath := os.Getenv("SNAP")
+		desktopFile := os.Getenv("BAMF_DESKTOP_FILE_HINT")
+		if snapPath != "" && desktopFile != "" {
+			// Only the application file name is required for launching.
+			desktopFile := filepath.Base(desktopFile)
+
+			// Attempt to launch the desktop file via the privileged launcher, this will check that this snap has the desktop-launch interface connected.
+			conn, err := dbusutil.SessionBusPrivate()
+			if err != nil {
+				fmt.Fprintf(Stderr, i18n.G("unable to access privileged desktop launcher: unable unable to get session bus: %v\n"), err)
+				os.Exit(46)
+			}
+			o := conn.Object("io.snapcraft.Launcher", "/io/snapcraft/PrivilegedDesktopLauncher")
+			call := o.Call("io.snapcraft.PrivilegedDesktopLauncher.OpenDesktopEntry", 0, desktopFile)
+			if call.Err != nil {
+				fmt.Fprintf(Stderr, i18n.G("failed to launch %s via the privileged desktop launcher: %v\n"), desktopFile, call.Err)
+				os.Exit(46)
+			}
+
+			os.Exit(0)
+		}
```

Comment:
I think it might make sense to move this code to `cmd_run.go` as part of the `cmdRun.Execute`, since that's the feature it is effectively overriding.